File: sillyhub-daemon/sillyhub_daemon/backends/jsonl.py
"""JsonlBackend — copilot JSONL dotted-event protocol backend.

Design reference: task-06, copilot.go.
Protocol: copilot CLI with ``--output-format json`` emits newline-delimited JSON.
Each line is ``{"type": "dotted.event.name", "data": {...}, ...}``.
"""

from __future__ import annotations

import asyncio
import json
import logging
import shutil
from dataclasses import dataclass, field
from typing import Any

from sillyhub_daemon.backends import AgentBackend, AgentEvent, TaskResult

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Internal state accumulator (mirrors copilotEventState in copilot.go)
# ---------------------------------------------------------------------------


@dataclass
class _JsonlState:
    """Mutable state accumulated while processing the JSONL event stream."""

    output: str = ""
    session_id: str = ""
    active_model: str = ""
    final_status: str = "completed"
    final_error: str = ""
    usage: dict[str, Any] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Backend implementation
# ---------------------------------------------------------------------------


class JsonlBackend(AgentBackend):
    """JSONL dotted-event protocol backend for copilot."""

    provider: str = "copilot"

    def __init__(self) -> None:
        self._state = _JsonlState()

    def _reset_state(self, seed_model: str = "") -> None:
        """Reset internal state for a fresh execution."""
        self._state = _JsonlState(active_model=seed_model or "copilot")

    # ── Argument builder ──────────────────────────────────────────────────

    def build_args(
        self,
        task_prompt: str,
        *,
        model: str = "",
        session_id: str = "",
    ) -> list[str]:
        """Assemble argv for a one-shot copilot invocation.

        Reference: ``buildCopilotArgs`` in copilot.go.

        .. code-block:: bash

            copilot -p "<prompt>" --output-format json --allow-all --no-ask-user
                    [--model <m>] [--resume <session-id>]
        """
        args: list[str] = [
            "-p",
            task_prompt,
            "--output-format",
            "json",
            "--allow-all",
            "--no-ask-user",
        ]
        if model:
            args.extend(["--model", model])
        if session_id:
            args.extend(["--resume", session_id])
        return args

    # ── Event parsing ────────────────────────────────────────────────────

    async def parse_output(self, line: str) -> AgentEvent | None:
        """Parse a single JSONL line into a structured event.

        Returns the *primary* event for the line. For event types that
        internally produce multiple logical events (e.g. ``assistant.message``
        with both reasoning and tool requests), only the first event is
        returned. Use ``parse_output_multi`` for full event extraction.
        """
        events = self.parse_output_multi(line)
        return events[0] if events else None

    def parse_output_multi(self, line: str) -> list[AgentEvent]:
        """Parse a single JSONL line, potentially returning multiple events.

        Used internally and by tests that need to inspect all events from
        a single line (e.g. ``assistant.message`` with reasoning + tool use).
        """
        line = line.strip()
        if not line:
            return []

        try:
            evt = json.loads(line)
        except (json.JSONDecodeError, ValueError):
            logger.warning("jsonl: failed to parse line: %s", line[:200])
            return []

        evt_type = evt.get("type", "")
        data = evt.get("data", {})

        return self._handle_event(evt_type, data, evt)

    def _handle_event(
        self,
        evt_type: str,
        data: dict[str, Any],
        raw: dict[str, Any],
    ) -> list[AgentEvent]:
        """Dispatch event by type, update state, return events.

        Reference: ``handleCopilotEvent`` in copilot.go.
        """
        events: list[AgentEvent] = []

        if evt_type == "session.start":
            self._handle_session_start(data)

        elif evt_type == "assistant.message_delta":
            ev = self._handle_message_delta(data)
            if ev:
                events.append(ev)

        elif evt_type == "assistant.message":
            events.extend(self._handle_message(data))

        elif evt_type in ("assistant.reasoning", "assistant.reasoning_delta"):
            ev = self._handle_reasoning(data)
            if ev:
                events.append(ev)

        elif evt_type == "tool.execution_complete":
            ev = self._handle_tool_complete(data)
            if ev:
                events.append(ev)

        elif evt_type == "assistant.turn_start":
            events.append(AgentEvent(event_type="status", status="running"))

        elif evt_type == "session.error":
            ev = self._handle_session_error(data)
            if ev:
                events.append(ev)

        elif evt_type == "session.warning":
            ev = self._handle_session_warning(data)
            if ev:
                events.append(ev)

        elif evt_type == "result":
            self._handle_result(raw)

        return events

    # ── Individual event handlers ─────────────────────────────────────────

    def _handle_session_start(self, data: dict[str, Any]) -> None:
        if data.get("selectedModel"):
            self._state.active_model = data["selectedModel"]
        if data.get("sessionId"):
            self._state.session_id = data["sessionId"]

    def _handle_message_delta(self, data: dict[str, Any]) -> AgentEvent | None:
        delta = data.get("deltaContent", "")
        if not delta:
            return None
        self._state.output += delta
        return AgentEvent(event_type="text", content=delta)

    def _handle_message(self, data: dict[str, Any]) -> list[AgentEvent]:
        """Handle assistant.message: reset output, emit reasoning/tool_use."""
        events: list[AgentEvent] = []

        content = data.get("content", "")
        if content:
            # Reset output to avoid delta double-counting.
            # Reference: copilot.go handleCopilotEvent assistant.message
            current = self._state.output
            if current.endswith(content):
                self._state.output = current[: -len(content)]
            # Add separator between turns if there's prior content
            if self._state.output and not self._state.output.endswith("\n\n"):
                self._state.output += "\n\n"
            self._state.output += content

        # Reasoning
        reasoning = data.get("reasoningText", "")
        if reasoning:
            events.append(AgentEvent(event_type="thinking", content=reasoning))

        # Tool requests
        tool_requests = data.get("toolRequests", [])
        for tr in tool_requests:
            tool_input: dict | None = None
            args = tr.get("arguments")
            if args:
                if isinstance(args, str):
                    try:
                        tool_input = json.loads(args)
                    except (json.JSONDecodeError, ValueError):
                        tool_input = {"raw": args}
                elif isinstance(args, dict):
                    tool_input = args
            events.append(
                AgentEvent(
                    event_type="tool_use",
                    tool_name=tr.get("name", ""),
                    call_id=tr.get("toolCallId", ""),
                    tool_input=tool_input,
                )
            )

        return events

    def _handle_reasoning(self, data: dict[str, Any]) -> AgentEvent | None:
        text = data.get("content", "") or data.get("deltaContent", "")
        if not text:
            return None
        return AgentEvent(event_type="thinking", content=text)

    def _handle_tool_complete(self, data: dict[str, Any]) -> AgentEvent | None:
        call_id = data.get("toolCallId", "")
        success = data.get("success", True)

        if success:
            result_obj = data.get("result")
            result_content = ""
            if result_obj and isinstance(result_obj, dict):
                result_content = result_obj.get("content", "")
        else:
            error_obj = data.get("error")
            if error_obj and isinstance(error_obj, dict):
                result_content = "Error: " + error_obj.get("message", "unknown")
            elif data.get("result") and isinstance(data["result"], dict):
                result_content = data["result"].get("content", "")
            else:
                result_content = "Error: unknown"

        return AgentEvent(
            event_type="tool_result",
            call_id=call_id,
            tool_output=result_content,
        )

    def _handle_session_error(self, data: dict[str, Any]) -> AgentEvent:
        msg = data.get("message", "unknown error")
        self._state.final_status = "failed"
        self._state.final_error = msg
        return AgentEvent(event_type="error", content=msg)

    def _handle_session_warning(self, data: dict[str, Any]) -> AgentEvent:
        msg = data.get("message", "")
        return AgentEvent(event_type="status", content=msg, level="warn")

    def _handle_result(self, raw: dict[str, Any]) -> None:
        if raw.get("sessionId"):
            self._state.session_id = raw["sessionId"]
        exit_code = raw.get("exitCode", 0)
        if exit_code != 0:
            self._state.final_status = "failed"
            exit_msg = f"copilot exited with code {exit_code}"
            if self._state.final_error:
                if exit_msg not in self._state.final_error:
                    self._state.final_error += "; " + exit_msg
            else:
                self._state.final_error = exit_msg

    # ── Execute ───────────────────────────────────────────────────────────

    async def execute(
        self,
        cmd_path: str,
        task_prompt: str,
        work_dir: str,
        env: dict | None = None,
        *,
        timeout: float = 0,
        model: str = "",
        session_id: str = "",
    ) -> TaskResult:
        """Execute copilot CLI and return structured result."""
        import time as _time

        seed_model = model or "copilot"
        self._reset_state(seed_model=seed_model)

        # Locate binary
        binary = cmd_path or "copilot"
        if not shutil.which(binary):
            return TaskResult(
                status="failed",
                output="",
                error=f"copilot executable not found: {binary}",
            )

        args = self.build_args(
            task_prompt,
            model=model,
            session_id=session_id,
        )

        start = _time.monotonic()
        collected_events: list[AgentEvent] = []

        try:
            proc = await asyncio.create_subprocess_exec(
                binary,
                *args,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=work_dir or None,
                env=env or None,
            )

            # Read stdout line by line
            assert proc.stdout is not None
            while True:
                raw_line = await proc.stdout.readline()
                if not raw_line:
                    break
                text_line = raw_line.decode(errors="replace")
                line_events = self.parse_output_multi(text_line)
                collected_events.extend(line_events)

            await proc.wait()

        except asyncio.TimeoutError:
            self._state.final_status = "timeout"
            self._state.final_error = f"copilot timed out after {timeout}s"
        except asyncio.CancelledError:
            self._state.final_status = "aborted"
            self._state.final_error = "execution cancelled"
        except Exception as exc:
            self._state.final_status = "failed"
            self._state.final_error = str(exc)

        duration_ms = int((_time.monotonic() - start) * 1000)

        return TaskResult(
            status=self._state.final_status,
            output=self._state.output,
            error=self._state.final_error,
            duration_ms=duration_ms,
            session_id=self._state.session_id,
            events=collected_events,
        )

File: sillyhub-daemon/src/adapters/jsonl.ts
/**
 * JsonlAdapter —— copilot CLI 的 JSONL 点分事件协议解析器。
 *
 * copilot CLI 启动参数 `--output-format json` 后，stdout 逐行输出 NDJSON，
 * 每行结构：`{"type": "dotted.event.name", "data": {...}, "sessionId"?, ...}`。
 * 本 adapter 把每行的点分 type 映射到统一 IR AgentEvent，并在实例字段
 * 维护 session 维度的累积状态（output 文本 / session_id / final_status）。
 *
 * 1:1 翻译自 Python `sillyhub_daemon/backends/jsonl.py`（parse_output_multi +
 * _handle_event + 8 个 _handle_* 子方法）。
 *
 * 方案B 定位（task-05）：本类只做纯解析，不负责子进程执行（spawn/stdin/
 * 超时）——执行下沉到 TaskRunner（task-19）。每个 lease 一个新实例，
 * 状态隔离，无需 reset。
 *
 * IR 收敛决策（与 task-02/06/07 全局一致）：
 *   - AgentEventType 为 5 元组（text/tool_use/tool_result/error/complete），
 *     无 status / thinking。
 *   - assistant.turn_start、session.warning → `type:'text' + metadata.status/level`
 *     （对齐 task-06 parseSystem/parseLog 的收敛方式）。
 *   - assistant.reasoning / reasoning_delta、assistant.message.reasoningText
 *     → `type:'text' + metadata.thinking:true`（对齐 task-06 thinking block）。
 *
 * metadata 字段名沿用 snake_case（types.ts L48-55 注释"保持 Python 原名以便
 * 对照调试"，与 task-06/07 全局一致）。
 *
 * @see design.md §7.1（AgentEvent IR）/ §7.3（PROTOCOL_PROVIDERS: jsonl→[copilot]）
 * @see Python 源 sillyhub_daemon/backends/jsonl.py
 */

import type { AgentEvent } from '../types.js';
import type { ProtocolAdapter } from './protocol-adapter.js';

/** jsonl adapter 内部累积状态（对应 Python _JsonlState，jsonl.py L28-36）。 */
interface JsonlState {
  output: string;
  sessionId: string;
  activeModel: string;
  finalStatus: 'completed' | 'failed' | 'timeout';
  finalError: string;
}

/**
 * copilot 的点分 JSONL 协议解析器。
 *
 * 一行可产出 0..N 个 AgentEvent：
 *   - 大多数 type 产出 0 或 1 个 event；
 *   - `assistant.message` 是唯一一行多 event 的 type（reasoning + 多个 tool_use）。
 *
 * 状态约定（task-05 B-03 允许有状态 adapter）：状态只在实例字段，不修改全局，
 * 不发起 I/O，不实现 control 应答方法（jsonl 协议无 stdin 应答需求，task-05 §B-02）。
 *
 * parse 返回 `AgentEvent[]`（不带 null）——空行 / 坏 JSON / 未知 type 一律返回 `[]`，
 * 与 Python `parse_output_multi` 永远返回 list 的语义一致（对照 task-06/07 返回
 * `| null` 的差异：jsonl 协议的「无事件」是合法且常见的，返回 [] 让消费端无需
 * 区分 null 与空数组，简化 TaskRunner 聚合逻辑）。
 */
export class JsonlAdapter implements ProtocolAdapter {
  readonly provider = 'copilot';

  private state: JsonlState = {
    output: '',
    sessionId: '',
    activeModel: '',
    finalStatus: 'completed',
    finalError: '',
  };

  /** 暴露只读 state 快照，供 TaskRunner（task-19）读取累积 output / session_id。 */
  getState(): Readonly<JsonlState> {
    return this.state;
  }

  /**
   * 解析一行 JSONL，返回 0..N 个 AgentEvent。
   *
   * 分发对照 Python `_handle_event`（jsonl.py L122-171）：按完整 type 字符串
   * switch（不拆分点分层级，理由见 B-04）。
   *
   * 不抛异常：坏行返回 []（task-05 §B-04），TaskRunner 另包 try-catch 兜底。
   */
  parse(line: string): AgentEvent[] {
    const trimmed = line.trim();
    if (!trimmed) return []; // B-05：空行 / 仅空白

    let evt: unknown;
    try {
      evt = JSON.parse(trimmed);
    } catch {
      return []; // B-02：坏 JSON 吞掉
    }
    if (!isRecord(evt)) return []; // 非对象

    const evtType = typeof evt.type === 'string' ? evt.type : '';
    const data = isRecord(evt.data) ? evt.data : {};

    switch (evtType) {
      case 'session.start':
        this.handleSessionStart(data);
        return [];

      case 'assistant.message_delta':
        return this.handleMessageDelta(data);

      case 'assistant.message':
        return this.handleMessage(data);

      case 'assistant.reasoning':
      case 'assistant.reasoning_delta':
        return this.handleReasoning(data);

      case 'tool.execution_complete':
        return this.handleToolComplete(data);

      case 'assistant.turn_start':
        // IR 收敛：status 合入 text + metadata.status（对齐 task-02/06）
        return [{ type: 'text', content: '', metadata: { status: 'running' } }];

      case 'session.error':
        return this.handleSessionError(data);

      case 'session.warning':
        return this.handleSessionWarning(data);

      case 'result':
        this.handleResult(evt);
        return [];

      default:
        // B-01：未知 type 静默丢弃（对齐 Python default 无 append）
        return [];
    }
  }

  // ── 各 type handler（对齐 Python _handle_* 方法）──────────────────────

  /** 对照 Python `_handle_session_start`（jsonl.py L175-179）。 */
  private handleSessionStart(data: Record<string, unknown>): void {
    const model = typeof data.selectedModel === 'string' ? data.selectedModel : '';
    const sessionId = typeof data.sessionId === 'string' ? data.sessionId : '';
    if (model) this.state.activeModel = model;
    if (sessionId) this.state.sessionId = sessionId;
  }

  /** 对照 Python `_handle_message_delta`（jsonl.py L181-186）。 */
  private handleMessageDelta(data: Record<string, unknown>): AgentEvent[] {
    const delta = typeof data.deltaContent === 'string' ? data.deltaContent : '';
    if (!delta) return [];
    this.state.output += delta;
    return [{ type: 'text', content: delta }];
  }

  /**
   * assistant.message：一行多 event 的核心场景。
   * 步骤（对齐 Python `_handle_message`，jsonl.py L188-231）：
   *   1. 若 content 非空：重置 output 防双计（先截尾、加 \n\n 分隔、append）；
   *   2. 若 reasoningText 非空：push thinking 事件（IR 收敛为 text + metadata.thinking）；
   *   3. 遍历 toolRequests[]：每个 push 一个 tool_use 事件（arguments 可为 string/dict）。
   */
  private handleMessage(data: Record<string, unknown>): AgentEvent[] {
    const events: AgentEvent[] = [];

    const content = typeof data.content === 'string' ? data.content : '';
    if (content) {
      // 防双计：若 output 已 endswith content（delta 先到），截掉尾部（B-07）。
      const current = this.state.output;
      if (current.endsWith(content)) {
        this.state.output = current.slice(0, -content.length);
      }
      // 加分隔符（若已有内容且不以 \n\n 结尾）。
      if (this.state.output && !this.state.output.endsWith('\n\n')) {
        this.state.output += '\n\n';
      }
      this.state.output += content;
    }

    // reasoning：IR 收敛为 text + metadata.thinking
    const reasoning = typeof data.reasoningText === 'string' ? data.reasoningText : '';
    if (reasoning) {
      events.push({ type: 'text', content: reasoning, metadata: { thinking: true } });
    }

    // tool requests（arguments 可为 string / dict，B-06）
    const toolRequests = Array.isArray(data.toolRequests) ? data.toolRequests : [];
    for (const tr of toolRequests) {
      if (!isRecord(tr)) continue;
      let toolInput: Record<string, unknown> | undefined;
      const args = tr.arguments;
      if (typeof args === 'string') {
        try {
          const parsed = JSON.parse(args);
          if (isRecord(parsed)) toolInput = parsed;
          else toolInput = { raw: args };
        } catch {
          toolInput = { raw: args };
        }
      } else if (isRecord(args)) {
        toolInput = args;
      }
      const name = typeof tr.name === 'string' ? tr.name : '';
      const callId = typeof tr.toolCallId === 'string' ? tr.toolCallId : '';
      events.push({
        type: 'tool_use',
        content: '',
        metadata: { tool_name: name, call_id: callId, tool_input: toolInput },
      });
    }

    return events;
  }

  /** 对照 Python `_handle_reasoning`（jsonl.py L233-237）。 */
  private handleReasoning(data: Record<string, unknown>): AgentEvent[] {
    // 对齐 Python `data.get("content", "") or data.get("deltaContent", "")`：
    // content 为空串（falsy）时 fallback 到 deltaContent，非「content 存在即取」。
    const content = typeof data.content === 'string' ? data.content : '';
    const deltaContent = typeof data.deltaContent === 'string' ? data.deltaContent : '';
    const text = content || deltaContent;
    if (!text) return [];
    return [{ type: 'text', content: text, metadata: { thinking: true } }];
  }

  /** 对照 Python `_handle_tool_complete`（jsonl.py L239-261）。 */
  private handleToolComplete(data: Record<string, unknown>): AgentEvent[] {
    const callId = typeof data.toolCallId === 'string' ? data.toolCallId : '';
    const success = data.success !== false; // 默认 true（对照 Python data.get("success", True)）
    let resultContent = '';

    if (success) {
      const resultObj = data.result;
      if (isRecord(resultObj)) {
        resultContent = typeof resultObj.content === 'string' ? resultObj.content : '';
      }
    } else {
      const errorObj = data.error;
      if (isRecord(errorObj)) {
        const msg = typeof errorObj.message === 'string' ? errorObj.message : 'unknown';
        resultContent = 'Error: ' + msg;
      } else if (isRecord(data.result)) {
        resultContent =
          typeof data.result.content === 'string' ? data.result.content : '';
      } else {
        resultContent = 'Error: unknown';
      }
    }

    return [
      {
        type: 'tool_result',
        content: resultContent,
        metadata: { call_id: callId, tool_output: resultContent },
      },
    ];
  }

  /** 对照 Python `_handle_session_error`（jsonl.py L263-267）。 */
  private handleSessionError(data: Record<string, unknown>): AgentEvent[] {
    const msg = typeof data.message === 'string' ? data.message : 'unknown error';
    this.state.finalStatus = 'failed';
    this.state.finalError = msg;
    return [{ type: 'error', content: msg }];
  }

  /**
   * 对照 Python `_handle_session_warning`（jsonl.py L269-271）。
   * IR 收敛：status 合入 text + metadata.level（对齐 task-06 parseLog）。
   */
  private handleSessionWarning(data: Record<string, unknown>): AgentEvent[] {
    const msg = typeof data.message === 'string' ? data.message : '';
    return [{ type: 'text', content: msg, metadata: { level: 'warn' } }];
  }

  /** 对照 Python `_handle_result`（jsonl.py L273-284）。 */
  private handleResult(raw: Record<string, unknown>): void {
    const sessionId = typeof raw.sessionId === 'string' ? raw.sessionId : '';
    if (sessionId) this.state.sessionId = sessionId;
    const exitCode = typeof raw.exitCode === 'number' ? raw.exitCode : 0;
    if (exitCode !== 0) {
      this.state.finalStatus = 'failed';
      const exitMsg = `copilot exited with code ${exitCode}`;
      if (this.state.finalError) {
        if (!this.state.finalError.includes(exitMsg)) {
          this.state.finalError += '; ' + exitMsg;
        }
      } else {
        this.state.finalError = exitMsg;
      }
    }
  }
}

// ─────────────────────────────────────────────────────────────────────────
// 私有 helper（模块级，纯函数）
// ─────────────────────────────────────────────────────────────────────────

/** 类型守卫：值是非 null 的 plain object（非数组）。 */
function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}
